res/controllers/Numark-N4-4channel-scripts.js

```diff
@@ -0,0 +1,674 @@
+      this.inKey="loop_out";
+      this.outKey="loop_out";
+    },
+  });
+  this.btnSample3 = new components.Button({
+    midi: [0x90+channel,0x15,0xB0+channel,0x0D],
+    shift: function () {
+      this.type=components.Button.prototype.types.toggle;
+      this.inKey="slip_enabled";
+      this.outKey="slip_enabled";
+    },
+    unshift: function () {
+      this.type=components.Button.prototype.types.push;
+      this.inKey="beatloop_activate";
+      this.outKey="beatloop_activate";
+    },
+  });
+  this.btnSample4 = new components.Button({
+    midi: [0x90+channel,0x16,0xB0+channel,0x0E],
+    shift: function () {
+      this.type=components.Button.prototype.types.toggle;
+      this.inKey="repeat";
+      this.outKey="repeat";
+    },
+    unshift: function () {
+      this.type=components.Button.prototype.types.push;
+      this.inKey="reloop_toggle";
+      this.outKey="loop_enabled";
+    },
+  });
+  // custom Hotcue Buttons
+  this.hotcueButtons=[];
+
+  for (var counter=0;counter<=3;counter++){
+    this.hotcueButtons[counter] = new components.HotcueButton({
+      midi: [0x90+channel,0x27+counter,0xB0+channel,0x18+counter],
+      number: counter+1,
+    });
+  }
+  this.encFxParam1 = new components.Encoder({
+    midi:[0xB0+channel,0x57],
+    group: "[EffectRack1_EffectUnit1]",
+    shift: function () {
+      this.inKey="mix";
+    },
+    unshift: function () {
+      this.inKey="super1";
+    },
+  });
+  this.encFxParam2 = new components.Encoder({
+    midi:[0xB0+channel,0x58],
+    group: "[EffectRack1_EffectUnit2]",
+    shift: function () {
+      this.inKey="mix";
+    },
+    unshift: function () {
+      this.inKey="super1";
+    },
+  });
+  this.encSample3 = new components.Encoder({
+    midi: [0xB0+channel,0x5A],
+    currentLoopSizeIndex: 7,
+    hotCuePage: 0,
+    applyHotcuePage: function (layer, displayFeedback) {
+      // ES3 doesn't allow default values in the function signature
+      // Could be replaced after migration to QJSEngine by "displayFeedback=true"
+      if (displayFeedback == undefined) {
+        displayFeedback = true;
+      }
+      layer = Math.max(Math.min(layer,3),0); // clamp layer value to [0;3] range
+      this.hotCuePage = layer;
+      if (this.timer) {engine.stopTimer(this.timer)}
+      var number = 0;
+      for (var i=0;i<theContainer.hotcueButtons.length;++i) {
+        number = (i+1)+theContainer.hotcueButtons.length*this.hotCuePage;
+        theContainer.hotcueButtons[i].disconnect();
+        theContainer.hotcueButtons[i].number=number;
+        theContainer.hotcueButtons[i].outKey='hotcue_' + number + '_enabled';
+        theContainer.hotcueButtons[i].unshift(); // for setting inKey based on number property.
+        theContainer.hotcueButtons[i].connect();
+        theContainer.hotcueButtons[i].trigger();
+      }
+      if (displayFeedback) {
+        for (var i=0;i<4;++i) {
+          midi.sendShortMsg(0xB0+channel,0x0B+i,(i-this.hotCuePage)?0x00:0x7F);
```

Comment:
The controller has a eight buttons arranged in two rows with four cells each. The second row is used to represent a set of four cuepoints. When you turn the encoder (which has hard stops which is why I use it as left/right buttons) you change the current "page" of hotcues. As it can be ease to loose track which page you are on, the buttons on the first row will briefly show which page you are currently on after you changed the page. So if I was on hotcue page one and turned the encoder one step to the left, all the light on second row would represent the new hotcue buttons and the second button of the first row would light up for 1s.
TLDR: It displays the current hotcuepage index within the upper row of the buttongrid.